### PR TITLE
Reimplementation of skfem.mesh

### LINKED ISSUE #503: Implement MeshHex.element_finder
After this `Basis.interpolator` would work with all mesh types.

### LINKED ISSUE #540: Steps towards a "functional Mesh"
We decided with @gdmcbain on the goal of having the public API of `Mesh` to be functional in the sense that none of the methods purposefully modify `self`. #536 and the corresponding PR #538 will eventually take care of `refine`, `scale`, `translate`, etc. by introducing `refined`, `scaled`, `translated` that return copies of the mesh. We will first use the new versions in the documentation and deprecate the old versions. Finally after a suitable time and a number of releases we will remove the original `refine`, `scale`, `translate`. There are immediate benefits to the functionality of `Mesh`, e.g., one cannot accidentally invalidate a `Basis` object by calling `Mesh.refine`.

However, there is a missing piece of the puzzle: `Mesh.define_boundary`. This was never such a huge success as one can simply provide `boundaries` dictionary to the constructor or then directly pass facet indices to `basis.get_dofs` or `basis.find_dofs`.

### LINKED ISSUE #541: Use Mesh.refined in docs, examples and tests
Now that `Mesh.refine` is deprecated it's time to get rid of it.

### LINKED ISSUE #546: Tasks before releasing 3.0.0
I expect 3.0.0 to be released before summer, so not yet. However, here is a checklist of things to do before releasing so that I won't forget:
- [x] finally delete `skfem.importers`
- [x] remove support of `lambda x, y, z` in `project`
- [x] remove `L2_projection`
- [x] remove `derivative`
- [x] remove the deprecated properties of `DiscreteField`
- [x] remove the deprecated `skfem.models.helpers`

### PR DESCRIPTION
**Pull request by kinnala**
- Added `MeshHex2`
- Added `MeshTet2`
- Added `MeshHex.to_meshtet`
- Added `MeshHex.element_finder` so that `Basis.interpolator` works for hexahedral meshes. Fixes #503 
- Removed `Mesh.{refine,scale,translate}`
- Removed `Mesh.define_boundary`
- Added `Mesh.with_boundaries` and `Mesh.with_subdomains`
- Changed `Mesh{Tri,Quad}.refined` to clear `subdomains` and `boundaries`, and not update them based on the new indexing (this will be hopefully reintroduced when resolving #271)
- Fixes #540
- Fixes #541
- Fixes #546

### CODE REVIEW

**Comment by kinnala** [comment]
Let me add that in theory it supports arbitrarily high order meshes.

**Comment by kinnala** [comment]
One thing that must be verified is that the performance cost is not too large because it's using `MappingIsoparametric` always in place of `MappingAffine`.

File: skfem/mesh/graph.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+        BOUNDARY_ELEMENT_MAP = {
+            ElementTriP1: ElementLineP1,
+            ElementTriP2: ElementLineP2,
+            ElementQuad1: ElementLineP1,
+            ElementQuad2: ElementLineP2,
+            ElementTetP1: ElementTriP1,
+            ElementTetP2: ElementTriP2,
+            ElementHex1: ElementQuad1,
+            ElementHex2: ElementQuad2,
+        }
```
Comment (gdmcbain) [inline skfem/mesh/graph.py:207]: Would these more neatly be attributes of the keys?

File: skfem/mesh/graph.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    @property
+    def _facet_indices(self):
+
+        if self.t.shape[0] == 3:
+            return [
+                [0, 1],
+                [1, 2],
+                [0, 2],
+            ]
+        elif self.t.shape[0] == 4:
```
[the remark this answers is not in the document]
Comment (gdmcbain) [reply skfem/mesh/graph.py:149]: Oh, is it just that quadrilaterals aren't included yet and will be handled by something like

https://github.com/kinnala/scikit-fem/blob/93c8bc189efee8090b49b90a2f6ab2fe5028a5a5/skfem/mapping/mapping_isoparametric.py#L202

[the remark this answers is not in the document]
Comment (gdmcbain) [reply skfem/mesh/graph.py:149]:
```python
[np.array_equal(m.facets, M.facets) for m, M in [(m, M), (m2, M2), (m3, M3), (m4, M4)]]
```
> [True, True, False, False]

**Review by gdmcbain — COMMENTED** [review]
```python
assert (a-A).sum() < 1e-10
```

Possibly not the best test since
```python
a.sum() == A.sum() == 0
```
and almost numerically also for `a2`, `A2`, &hellip;, by definition the laplacian of a constant vanishing.

Maybe 
```python
assert (a - A).nnz == (a3 - A3).nnz == (a4 - A4).nnz == 0
```
but `(a2 - A2).nnz == 452` and has data ranging from -504.5 to 251.6.

File: skfem/mesh/graph.py (hunk 1 above)
Comment (kinnala) [reply skfem/mesh/graph.py:207]: Yeah it makes sense.

File: skfem/mesh/graph.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kinnala) [reply skfem/mesh/graph.py:149]: Thanks for pointing these out. It seems that the quadratic Grid still has some issues. It's a bit complicated to wrap my head around it, however, at least we have the working reference `MeshTri2` to compare against.

**Comment by kinnala** [comment]
- [ ] Where should `.refined` etc. go?
- [ ] Should `MappingIsoparametric` be part of `Grid` going forwards?

**Comment by kinnala** [comment]
Although it's very nice approach, currently the performance difference is so large that I'll dump it unless I'm able to optimize it:
```
In [1]: %timeit basis3 = InteriorBasis(m3, ElementTetP1())                      
142 ms ± 773 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [2]: %timeit Basis3 = InteriorBasis(M3, ElementTetP1())                      
775 ms ± 5.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
There are some bottlenecks that I immediately recognize. I'll start optimizing it tomorrow.

**Comment by kinnala** [comment]
Tests pass for first time in two weeks! Now I think I have solved all the technical challenges (i.e. how to represent the mesh and how to load from external formats) but there is still a fair bit of refactoring and testing to do before it can be merged. This will most likely drop support for Python 3.6 if we want to use dataclasses. Another options are to use a backported version of dataclasses or implement some constructors and other things ourselves before Python 3.6 support is officially dropped.

**Comment by kinnala** [comment]
Currently the diff is +300 LOC but when this is ready I expect to be able to remove the entire old `skfem.mesh` module.

**Comment by kinnala** [comment]
@gdmcbain Now there are less than 10 of the original tests failing. Soon I'll start refactoring and cleaning it up. This will definitely now be 3.0.0. Some missing work:
- [ ] Reimplement MeshLine
- [ ] Implement MeshHex2 and MeshTet2
- [ ] Subdomains/boundaries in refinement (#104)
- [ ] Explicit invalidate in remaining cases (#542)
- [ ] Element finder for MeshHex1 (#503)
- [ ] Update element finder for MeshTri (currently depends on matplotlib; reimplement in scipy.ckdtree instead)
- [ ] Look into #411 and #412; former should be now easy, also plan is to have some sort of basic support for #412 without hacking `basis.N` as demonstrated in the comments of #412
- [ ] #261; I plan to deprecate our in-house JSON format and try to use meshio instead
- [ ] Possibly look into #105 just to get more features to 3.0.0.
- [ ] Documentation of BaseMesh and subclasses, and update of tests (#541)
- [ ] Documentation of RefDom
- [ ] Compare performance (I've done some preliminary tests and everything should be now fine since it's still using the original Mapping classes; next up is the reimplementation of Mapping but not for this PR.)
- [ ] Changelog entries

**Comment by kinnala** [comment]
I wonder why these tests won't run...

**Comment by gdmcbain** [comment]
> I wonder why these tests won't run...

I don't know either; however, they do run on my branch and there are ten failures there too.   https://github.com/kinnala/scikit-fem/compare/next-gen-mesh...gdmcbain:next-gen-mesh

I fixed a couple (exx 11, 13) using `Mesh.translated` instead of mutating `.p`, but those were low-hanging fruit.  Also I suspect that there might be a better way to do ex13 than with `Mesh.translated` once I look more closely at all this very enticing new work.

**Comment by kinnala** [comment]
One test failing. Some tests are disabled for now, e.g., mesh validity checks as I haven't still decided what to do with those. Maybe making it a separate method which is not running automatically?

**Comment by kinnala** [comment]
Zero tests failing (mypy fails).

**Comment by kinnala** [comment]
🍾🍾🍾
I think I'll want to merge this ASAP after some further tests and cleaning up.

Some thoughts:
- `define_boundary` is still there to prove backwards-compatibility. [ ] Make it return a new mesh instead.
- [ ] Implement MeshHex2 and MeshTet2
- [ ] Element finder for MeshHex1 (#503)
- [ ] Update element finder for MeshTri (currently depends on matplotlib; reimplement in scipy.ckdtree instead)

**Comment by kinnala** [comment]
TODO hexahedron save reordering

**Pull request merged**